### ISSUE TITLE
Detail view not unmounted correctly in People demo

### ISSUE THREAD

**Issue by Almo7aya** [developit/preact#1475]
I think the newly added demo is not working correctly, the remove contact button doesn't remove the contact profile!?

**Comment by Almo7aya** [comment]
They stack one on one 
![Capture](https://user-images.githubusercontent.com/17254073/55224434-980e8e80-5221-11e9-8de2-a74c386f7586.PNG)

**Comment by marvinhagemeister** [comment]
The people demo isn't broken. We specifically added it because it is a very good real world app that relies on several popular third-party libraries like `mobx` and `mobx-state-tree`. The demo works with React so if anything is broken it has something with Preact. The point of the demo is to catch such issues.

**Comment by Almo7aya** [comment]
I got your point. There is a weird thing going on, the demo adds a new div every time I click a user. 
![Capture](https://user-images.githubusercontent.com/17254073/55228292-d65c7b80-522a-11e9-8d70-376bc60b8424.PNG)

**Comment by marvinhagemeister** [comment]
Had a look into this and it's caused by a mismatch of the DOM reference on a `component`. We somehow end up in a state where `component.base` points to the correct element, but `component._vnode._dom` does not. The latter is `null` instead which leads to `diff()` assuming that a new `<div>` element needs to be created.

https://github.com/developit/preact/blob/d63b740a8a471758b9243aa083e61bc3484fa6c4/src/component.js#L65

Replacing `dom = this._vnode._dom` with `dom = this.base` fixes the issue, but I'm unable to reproduce it with a test case so far. Haven't understood yet what's causing the inconsistency.

**Comment by Almo7aya** [comment]
Returning a `<div/>` instend of null here fixes this issue
https://github.com/developit/preact/blob/d63b740a8a471758b9243aa083e61bc3484fa6c4/demo/people/profile.tsx#L23


But it creates more `div`s with also `class=profile`

![Capture](https://user-images.githubusercontent.com/17254073/55232576-090c7100-5237-11e9-80de-0fa89c904b9f.PNG)

**Comment by marvinhagemeister** [comment]
Removing the `Route` components also works. My gut tells me that this could be related to `Fragment` reconciliation, but I don't know yet for sure.

**Comment by marvinhagemeister** [comment]
I think I got one. It's super weird but somehow props is not even updated :man_shrugging: 

```jsx
it('todo description', () => {
	let bar;
	class Bar extends Component {
		constructor(props) {
			super(props);
			bar = this;
		}

		render() {
			return <Foo />;
		}
	}

	let foo;
	class Foo extends Component {
		constructor(props) {
			super(props);
			foo = this;
		}

		shouldComponentUpdate() {
			return false;
		}

		render() {
			if (this.props.hidden) return null;
			return <div>foo</div>;
		}
	}

	render(<Bar />, scratch);
	expect(scratch.textContent).to.equal('foo');

	foo.props.hidden = true;
	foo.forceUpdate();
	bar.forceUpdate();

	expect(foo.props).to.deep.equal({ hidden: true });
	expect(scratch.textContent).to.equal('');
});
```

**Comment by marvinhagemeister** [comment]
Fixed thanks to #1484 and #1489 :tada:
